File: sigil-hs/src/Sigil/Codec/ColorTransform.hs
module Sigil.Codec.ColorTransform
  ( forwardRCT
  , inverseRCT
  ) where

import Data.Int  (Int32)
import Data.Word (Word8)
import qualified Data.Vector.Unboxed as VU

-- | Forward Reversible Color Transform.
--
-- Takes width, height, and interleaved RGB pixel data (width * height * 3 bytes).
-- Returns three separate channel arrays of Int32: (Y, Cb, Cr).
--
-- For grayscale (1 channel): just converts Word8 -> Int32.
-- For RGBA (4 channels): applies RCT to RGB, passes alpha separately.
forwardRCT :: Int -> Int -> VU.Vector Word8 -> (VU.Vector Int32, VU.Vector Int32, VU.Vector Int32)
forwardRCT w h pixels
  | npx == 0  = (VU.empty, VU.empty, VU.empty)
  | otherwise = (yChannel, cbChannel, crChannel)
  where
    npx = w * h
    yChannel  = VU.generate npx $ \i ->
      let r = fromIntegral (pixels VU.! (i * 3))     :: Int32
          g = fromIntegral (pixels VU.! (i * 3 + 1)) :: Int32
          b = fromIntegral (pixels VU.! (i * 3 + 2)) :: Int32
      in (r + 2 * g + b) `div` 4
    cbChannel = VU.generate npx $ \i ->
      let g = fromIntegral (pixels VU.! (i * 3 + 1)) :: Int32
          b = fromIntegral (pixels VU.! (i * 3 + 2)) :: Int32
      in b - g
    crChannel = VU.generate npx $ \i ->
      let r = fromIntegral (pixels VU.! (i * 3))     :: Int32
          g = fromIntegral (pixels VU.! (i * 3 + 1)) :: Int32
      in r - g

-- | Inverse Reversible Color Transform.
--
-- Takes width, height, and three channel arrays (Y, Cb, Cr).
-- Returns interleaved RGB pixel data as Word8.
inverseRCT :: Int -> Int -> (VU.Vector Int32, VU.Vector Int32, VU.Vector Int32) -> VU.Vector Word8
inverseRCT w h (yChannel, cbChannel, crChannel)
  | npx == 0  = VU.empty
  | otherwise = VU.generate (npx * 3) $ \idx ->
      let i = idx `div` 3
          c = idx `mod` 3
          yr = yChannel  VU.! i
          cb = cbChannel VU.! i
          cr = crChannel VU.! i
          g  = yr - (cb + cr) `div` 4
          r  = cr + g
          b  = cb + g
      in case c of
           0 -> clampWord8 r
           1 -> clampWord8 g
           _ -> clampWord8 b
  where
    npx = w * h

clampWord8 :: Int32 -> Word8
clampWord8 x
  | x < 0    = 0
  | x > 255  = 255
  | otherwise = fromIntegral x

File: sigil-rs/src/color_transform.rs
/// Inverse Reversible Color Transform (RCT).
///
/// Converts (Y, Cb, Cr) channel arrays back to interleaved RGB bytes.
/// Matches the Haskell `inverseRCT` in ColorTransform.hs:
///   g = Y  - (Cb + Cr) `div` 4
///   r = Cr + g
///   b = Cb + g
pub fn inverse_rct(width: usize, height: usize, y: &[i32], cb: &[i32], cr: &[i32]) -> Vec<u8> {
    let n = width * height;
    let mut pixels = Vec::with_capacity(n * 3);
    for i in 0..n {
        let g = y[i] - (cb[i] + cr[i]).div_euclid(4);
        let r = cr[i] + g;
        let b = cb[i] + g;
        pixels.push(r.clamp(0, 255) as u8);
        pixels.push(g.clamp(0, 255) as u8);
        pixels.push(b.clamp(0, 255) as u8);
    }
    pixels
}

/// Clamp an i32 to [0, 255] and cast to u8.
pub fn clamp_u8(v: i32) -> u8 {
    v.clamp(0, 255) as u8
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_inverse_rct_white() {
        // White pixel: R=255, G=255, B=255
        // Forward RCT: Y = (255+2*255+255)/4 = 255, Cb = 255-255 = 0, Cr = 255-255 = 0
        let y  = vec![255i32];
        let cb = vec![0i32];
        let cr = vec![0i32];
        let pixels = inverse_rct(1, 1, &y, &cb, &cr);
        assert_eq!(pixels, vec![255, 255, 255]);
    }

    #[test]
    fn test_inverse_rct_red() {
        // Red pixel: R=255, G=0, B=0
        // Forward RCT: Y = (255+0+0)/4 = 63, Cb = 0-0 = 0, Cr = 255-0 = 255
        // Inverse: g = 63 - (0+255)/4 = 63 - 63 = 0, r = 255+0 = 255, b = 0+0 = 0
        let y  = vec![63i32];
        let cb = vec![0i32];
        let cr = vec![255i32];
        let pixels = inverse_rct(1, 1, &y, &cb, &cr);
        assert_eq!(pixels, vec![255, 0, 0]);
    }

    #[test]
    fn test_inverse_rct_clamp() {
        // Values that would overflow — should be clamped to [0, 255]
        let y  = vec![300i32];
        let cb = vec![0i32];
        let cr = vec![0i32];
        let pixels = inverse_rct(1, 1, &y, &cb, &cr);
        assert_eq!(pixels, vec![255, 255, 255]);
    }
}
